C# 6.0 feature: Exception Filters

BillWagner opened the pull request:
Add "when" as a contextual keyword.
Update grammar for exception filters
Add description of exception filters.

BillWagner committed:
feature exception filters

Add "when" as a contextual keyword.
Update grammar for exception filters
Add description of exception filters.

RexJaeschke commented:
I originally created this proposal in the conversion-to-markdown repo as [PR107](https://github.com/ECMA-TC49-TG2/conversion-to-markdown/pull/107). It had two commits, to update statements.md and lexical-structure.md. Later, @BillWagner force-pushed that work over here, but as best as I can tell, the front part of the statements.md commit got lost, and needs to be put back. (Bill, being a novice GitHub user, rather than trying to figure our how to salvage that missing text from the commit, I'm showing it [the complete/replacement *try_statement* grammar] here, so Nigel can continue his review.)

```antlr
try_statement
    : 'try' block catch_clause+
    | 'try' block finally_clause
    | 'try' block catch_clause+ finally_clause
    ;
catch_clause
    : 'catch' exception_specifier? exception_filter?  block
    ;
exception_specifier
    : '(' type identifier? ')'
    ;
exception_filter
    : 'when' '(' expression ')'
    ;
finally_clause
    : 'finally' block
    ;
```

@BillWagner  - I updated this PR with the above grammar on 2021/07/15

Nigel-Ecma commented:
Thanks for the proposed grammar Rex, makes it easier :-) 

An observation on the grammar, I'll decide later whether it is more than that (i.e. this is my note pad!)

For comparison this is the current grammar in ANTLR:

```ANTLR
try_statement
    : 'try' block catch_clauses
    | 'try' block catch_clauses* finally_clause
    ;
catch_clauses
    : specific_catch_clause+
    | specific_catch_clause* general_catch_clause
    ;
specific_catch_clause
    : 'catch' '(' type identifier? ')'  block
    ;
general_catch_clause
    : 'catch' block
    ;
finally_clause
    : 'finally' block
    ;
```

Notes:

- In the original grammar *try_statement* syntactically enforces that there must be at least one `catch` or `finally` in a `try`, and `finally` must be last if there is one.
- Similarly *catch_clauses* syntactically enforces that a *general_catch_clause* must be last
- In the proposed grammar *try_statement* is expanded to three alternatives (the syntax is equivalent to the original), maybe to match the text’s “There are three possible forms...”?, and continues to syntactically enforce the same conditions as the original grammar
- However that a catch clause without any *exception_specifier* or *exception_filter* (the original’s *general_catch_clause*) must be last is no longer enforced syntactically and moves to the text as a semantic rule.

Conclusion? None, for now just an observation...

Nigel-Ecma commented on standard/statements.md:
```diff
@@ -1103,17 +1103,17 @@ There are three possible forms of `try` statements:
 - A `try` block followed by a `finally` block.
 - A `try` block followed by one or more `catch` blocks followed by a `finally` block.
```
This original para & bullet list are redundant, the forms are not referenced elsewhere and it simply a prose description of the `try_statement` production. In the original that production only has two alternatives:

```
try_statement
    : 'try' block catch_clauses
    | 'try' block catch_clauses* finally_clause
    ;
```

while the proposal expands this to three, maybe just to match this (original) bullet list, though the described syntax is identical.

The descriptions of other statements do not generally start with a prose description of the statement structure, so there is not a pattern being followed here, however the switch statement does.

Option 1: 😄 Remove this para & bullet list

Option 2: 🚀 Follow the model of the switch statement and replaced with a more succinct description which doesn't introduce "forms". Something like:

> A *try_statement* consists of the keyword `try` followed by a *block*, then zero or more *catch_clauses*, then an optional *finally_clause*. There must be at least one *catch_clause* or a *finally_clause*.

Option 3: 👀 Leave as is

If Option 1 or 2 is chosen then the grammar for *try_statement* can revert to the two alternative original version, of option 3 is chosen then there is a small argument for the proposed three alternative version.

BillWagner commented on standard/statements.md:
```diff
@@ -1103,17 +1103,17 @@ There are three possible forms of `try` statements:
 - A `try` block followed by one or more `catch` blocks followed by a `finally` block.
 
-When a `catch` clause specifies a *type*, the type shall be `System.Exception` or a type that derives from `System.Exception`. When a `catch` clause specifies a *type_parameter* it shall be a type parameter type whose effective base class is or derives from `System.Exception`.
+When a `catch` clause specifies an *exception_specifier*, the type shall be `System.Exception` or a type that derives from `System.Exception`.
```
Done.

BillWagner commented on standard/statements.md:
```diff
@@ -1103,17 +1103,17 @@ There are three possible forms of `try` statements:
+Unless a `catch` clause includes an exception variable name, it is impossible to access the exception object in the filter and `catch` block.
 
 A `catch` clause that specifies neither an exception type nor an exception variable name is called a general `catch` clause. A `try` statement can only have one general `catch` clause, and, if one is present, it shall be the last `catch` clause.
 
 > *Note*: Some programming languages might support exceptions that are not representable as an object derived from `System.Exception`, although such exceptions could never be generated by C# code. A general `catch` clause might be used to catch such exceptions. Thus, a general `catch` clause is semantically different from one that specifies the type `System.Exception`, in that the former might al ...
 
-In order to locate a handler for an exception, `catch` clauses are examined in lexical order. A compile-time error occurs if a `catch` clause specifies a type that is the same as, or is derived from, a type that was specified in an earlier `catch` clause for the same try.
+In order to locate a handler for an exception, `catch` clauses are examined in lexical order. If a `catch` clause specifies a type but no exception filter, it is a compile-time error for a later `catch` clause in the same `try` statement to specify a type that is the same as, or is derived from, that type. If a `catch` clause specifies no type and no filter, it shall be the last `catch` clause fo ...
```
Line numbers moved. The above comment references 1122 and 1116 now instead of 1116 and 1112, respectively.  I made the edit.

BillWagner committed:
respond to suggestions

BillWagner committed:
restore grammar updates

BillWagner commented:
Status update:

Most of the comments above have been addressed.  The following are still unresolved:

- Changes arising from https://github.com/dotnet/csharpstandard/pull/2#issuecomment-744665141.  I don't see a parallel PR, so we could discuss Nigel's suggested bullets or add a prose description to this PR.
- Nigel's set of proposals in https://github.com/dotnet/csharpstandard/pull/2#discussion_r542074435.  (My own preference is option 2).

gafter commented on standard/statements.md:
```diff
@@ -1109,17 +1113,17 @@ There are three possible forms of `try` statements:
 > *Note*: Some programming languages might support exceptions that are not representable as an object derived from `System.Exception`, although such exceptions could never be generated by C# code. A general `catch` clause might be used to catch such exceptions. Thus, a general `catch` clause is semantically different from one that specifies the type `System.Exception`, in that the former might al ...
 
-In order to locate a handler for an exception, `catch` clauses are examined in lexical order. A compile-time error occurs if a `catch` clause specifies a type that is the same as, or is derived from, a type that was specified in an earlier `catch` clause for the same try.
+In order to locate a handler for an exception, `catch` clauses are examined in lexical order. If a `catch` clause specifies a type but no exception filter, it is a compile-time error for a later `catch` clause in the same `try` statement to specify a type that is the same as, or is derived from, that type.
```
```suggestion
In order to locate a handler for an exception, `catch` clauses are examined in lexical order. If a `catch` clause specifies a type but no exception filter, it is a compile-time error for a later `catch` clause of the same `try` statement to specify a type that is the same as, or is derived from, that type.
```

gafter commented on standard/statements.md:
```diff
@@ -1176,9 +1180,10 @@ A `try` statement is executed as follows:
   - Control is transferred to the end point of the `try` statement.
 - If an exception is propagated to the `try` statement during execution of the `try` block:
-  - The `catch` clauses, if any, are examined in order of appearance to locate a suitable handler for the exception. The first `catch` clause that specifies the exception type or a base type of the exception type is considered a match. A general `catch` clause is considered a match for any exception type. If a matching `catch` clause is located:
+  - The `catch` clauses, if any, are examined in order of appearance to locate a suitable handler for the exception. If a `catch` clause does not specify a type, or specifies the exception type or a base type of the exception type is considered a match. A general `catch` clause is considered a match for any exception type. If a matching `catch` clause is located:
```
```suggestion
  - The `catch` clauses, if any, are examined in order of appearance to locate a suitable handler for the exception. Each `catch` clause that does not specify a type, or specifies the exception type or a base type of the exception type, is considered a match. If a matching `catch` clause is located:
```

gafter replied:
Remove duplicate description of general catch clause.

gafter commented on standard/statements.md:
```diff
@@ -1176,9 +1180,10 @@ A `try` statement is executed as follows:
+  - The `catch` clauses, if any, are examined in order of appearance to locate a suitable handler for the exception. If a `catch` clause does not specify a type, or specifies the exception type or a base type of the exception type is considered a match. A general `catch` clause is considered a match for any exception type. If a matching `catch` clause is located:
     - If the matching `catch` clause declares an exception variable, the exception object is assigned to the exception variable.
-    - Control is transferred to the matching `catch` block.
+    - If the `catch` clause declares an exception filter, the filter is evaluated. If it evaluates to `false`, the catch clause is not a match, and the search continues through any subsequent `catch` clauses for a suitable handler.    
```
```suggestion
    - If the `catch` clause declares an exception filter, the filter is evaluated. If it evaluates to `false`, the search continues through any subsequent `catch` clauses for a suitable handler.    
```

gafter replied:
Remove text that asserts that a catch clause that is a match is not a match.

gafter commented on standard/statements.md:
```diff
@@ -1176,9 +1180,10 @@ A `try` statement is executed as follows:
     - If the matching `catch` clause declares an exception variable, the exception object is assigned to the exception variable.
-    - Control is transferred to the matching `catch` block.
+    - If the `catch` clause declares an exception filter, the filter is evaluated. If it evaluates to `false`, the catch clause is not a match, and the search continues through any subsequent `catch` clauses for a suitable handler.    
+    - Otherwise, the `catch` clause is considered a match, and control is transferred to the matching `catch` block.
```
```suggestion
    - Otherwise control is transferred to the corresponding `catch` block.
```

gafter reviewed (dismissed):
Some mall changes suggested.

KalleOlaviNiemitalo commented on standard/statements.md:
```diff
@@ -1109,17 +1113,17 @@ There are three possible forms of `try` statements:
+In an *exception_specifier* the *type*, or its effective base class if it is a *type_parameter*, shall be `System.Exception` or a type that derives from it.
 
-When a `catch` clause specifies both a *class_type* and an *Identifier*, an ***exception variable*** of the given name and type is declared. The exception variable corresponds to a local variable with a scope that extends over the `catch` block. During execution of the `catch` block, the exception variable represents the exception currently being handled. For purposes of definite assignment check ...
+When a `catch` clause specifies both a *class_type* and an *Identifier*, an ***exception variable*** of the given name and type is declared. The exception variable corresponds to a local variable with a scope that extends over the `catch` block. During execution of the *exception_filter* and `catch` block, the exception variable represents the exception currently being handled. For purposes of de ...
```
It seems odd that the exception is already "being handled" during execution of the *exception_filter* even though an "exception handler" has not been located yet.

BillWagner committed:
Apply suggestions from code review

Co-authored-by: Neal Gafter <neal@gafter.com>

Nigel-Ecma commented:
Back in the mists of time (Dec 20) I remarked that the revised grammar which adds exception filters (`when`) also *removed* the syntactic requirement that a *general_catch_clause* if present must be the last one; leaving it as an in-text requirement only.

Being of the view that a *reference* grammar should be as precise as reasonable (while a compiler may use a more lenient grammar to assist error messages and recovery) here is a version of the grammar which maintains the syntactic restriction:

```ANTLR
try_statement
    : 'try' block catch_clauses
    | 'try' block catch_clauses* finally_clause
    ;
catch_clauses
    : specific_catch_clause+
    | specific_catch_clause* general_catch_clause
    ;
specific_catch_clause
    : 'catch' exception_specifier exception_filter? block
    : 'catch' exception_filter block
    ;
exception_specifier
    : '(' type identifier? ')'
    ;
exception_filter
    : 'when' '(' boolean_expression ')'
    ;
general_catch_clause
    : 'catch' block
    ;
finally_clause
    : 'finally' block
    ;
```

Notes:

1. I left the origianal 2-alternative version of *try_statement* rather than replace the the equivalent 3-alternative version
2. I changed *expression* to *boolean_expression* in *exception_filter* as that matches the purpose of *boolean_expression*. A matching change to the list of places *boolean_expression* occurs would be needed to *§12.21 Boolean expressions*.


I’ll argue this version of the grammar is better as a reference grammar as it embodies all of the syntactic requirements.

(And if case someone asks, sorry I haven't come up with a series of short paras to replace the nested bullets, over to a wordsmith ;-))

BillWagner committed:
Merge branch 'draft-v6' into feature-exception-filter

jskeet commented:
Recording my vote here for option 2 described above - a succinct description.

jskeet commented:
Further actions:

- Grammar as per https://github.com/dotnet/csharpstandard/pull/2#issuecomment-922606602
- Option 2 as per https://github.com/dotnet/csharpstandard/pull/2#discussion_r542074435
- Raise issue to potentially change nested bullet list to paragraphs, as per https://github.com/dotnet/csharpstandard/pull/2#issuecomment-744665141

jskeet reviewed (approved):
Approved after the grammar and "option 2" changes have been applied.

BillWagner committed:
Update recommended at meeting

Update the grammar and the description of the `try` statement.  See https://github.com/dotnet/csharpstandard/pull/2#issuecomment-948018076
